src/extensions/rate_limit/mod.rs

```diff
@@ -34,137 +50,81 @@ fn default_jitter_up_to_millis() -> u64 {
 
 pub struct RateLimitBuilder {
     config: RateLimitConfig,
+    ip_jitter: Option<Jitter>,
+    ip_limiter: Option<Arc<DefaultKeyedRateLimiter<String>>>,
 }
 
 #[async_trait::async_trait]
 impl Extension for RateLimitBuilder {
     type Config = RateLimitConfig;
 
     async fn from_config(config: &Self::Config, _registry: &ExtensionRegistry) -> Result<Self, anyhow::Error> {
-        Ok(Self::new(*config))
+        Ok(Self::new(config.clone()))
     }
 }
 
 impl RateLimitBuilder {
     pub fn new(config: RateLimitConfig) -> Self {
-        assert!(config.burst > 0, "burst must be greater than 0");
-        assert!(config.period_secs > 0, "period_secs must be greater than 0");
-        Self { config }
-    }
-    pub fn build(&self) -> RateLimit {
-        let burst = NonZeroU32::new(self.config.burst).unwrap();
-        let period = Duration::from_secs(self.config.period_secs);
-        let jitter = Jitter::up_to(Duration::from_millis(self.config.jitter_up_to_millis));
-        RateLimit::new(burst, period, jitter)
-    }
-}
-
-#[derive(Clone)]
-pub struct RateLimit {
-    burst: NonZeroU32,
-    period: Duration,
-    jitter: Jitter,
-}
-
-impl RateLimit {
-    pub fn new(burst: NonZeroU32, period: Duration, jitter: Jitter) -> Self {
-        Self { burst, period, jitter }
-    }
-}
-
-impl<S> tower::Layer<S> for RateLimit {
-    type Service = ConnectionRateLimit<S>;
-
-    fn layer(&self, service: S) -> Self::Service {
-        ConnectionRateLimit::new(service, self.burst, self.period, self.jitter)
-    }
-}
-
-#[derive(Clone)]
-pub struct ConnectionRateLimit<S> {
-    service: S,
-    limiter: Arc<DefaultDirectRateLimiter>,
-    jitter: Jitter,
-}
-
-impl<S> ConnectionRateLimit<S> {
-    pub fn new(service: S, burst: NonZeroU32, period: Duration, jitter: Jitter) -> Self {
-        let replenish_interval_ns = period.as_nanos() / (burst.get() as u128);
-        let quota = Quota::with_period(Duration::from_nanos(replenish_interval_ns as u64))
-            .unwrap()
-            .allow_burst(burst);
-        let limiter = Arc::new(RateLimiter::direct(quota));
-        Self {
-            service,
-            limiter,
-            jitter,
+        // make sure there is at least one rule
+        assert!(!config.rules.is_empty(), "must have at least one rule");
+        // make sure there is at most one ip rule
+        assert!(
```

Comment:
we might as well change the config to something like 
```yml
  rate_limit:
     connection:
       burst: 20 # 20 RPC requests per second per connection
       period_secs: 1
    ip:
       burst: 500 # 500 RPC requests per 10 seconds per ip
       period_secs: 10
```

and then we don't need extra validations for rule count

Reply:
Yeah makes sense